[Version 9.0] Feature support for `init` accessors

BillWagner opened the pull request:
This PR contains the work for `init` accessors in C# 9.

The commits from #978 were squashed to one commit in this branch.

There are a number of comments that haven't been addressed on #978:

- https://github.com/dotnet/csharpstandard/pull/978/files#r2326611994
- https://github.com/dotnet/csharpstandard/pull/978/files#r1378416566
- https://github.com/dotnet/csharpstandard/pull/978/files#r2326613914
- 

Notes:

- The ANTLR grammar validation fails for one test where the validation is designed to fail – it is testing that right shift is not erroneously recognised – and the checker passes the test if the error message is what it expected. Antlr writes a long informative error message listing all the tokens it might have expected to find at the failing point of the code and that list now has `init` in it.
- The renumbering action fails because this feature depends on the grammar in "records".

RexJaeschke committed:
Add files via upload

Add support for init accessors

Add support for init accessors

Add support for init accessors

Add support for init accessors

Add support for init accessors

Add support for init accessors

Add support for init accessors

fix md formatting

fix formatting, add xref links

add xref links

RexJaeschke committed:
Add files via upload

Add support for init accessors

Add support for init accessors

Add support for init accessors

Add support for init accessors

Add support for init accessors

Add support for init accessors

Add support for init accessors

fix md formatting

fix formatting, add xref links

add xref links

BillWagner committed:
fix word converter issues

Use different reasonable placeholders.

BillWagner committed:
fix word converter issues

Use different reasonable placeholders.

RexJaeschke commented on standard/basic-concepts.md:
```diff
@@ -81,7 +81,7 @@ There are several different types of declaration spaces, as described in the fol
 - Each delegate declaration creates a new declaration space. Names are introduced into this declaration space through parameters (*fixed_parameter*s and *parameter_array*s) and *type_parameter*s.
 - Each enumeration declaration creates a new declaration space. Names are introduced into this declaration space through *enum_member_declarations*.
-- Each method declaration, property declaration, property accessor declaration, indexer declaration, indexer accessor declaration, operator declaration, instance constructor declaration, anonymous function, and local function creates a new declaration space called a ***local variable declaration space***. Names are introduced into this declaration space through parameters (*fixed_parameter*s and  ...
+- Each method declaration, property declaration, property accessor declaration, indexer declaration, indexer accessor declaration, operator declaration, instance constructor declaration, anonymous function, and local function creates a new declaration space called a ***local variable declaration space***. Names are introduced into this declaration space through parameters (*fixed_parameter*s and  ...
```
```suggestion
- Each method declaration, property declaration, property accessor declaration, indexer declaration, indexer accessor declaration, operator declaration, instance constructor declaration, anonymous function, and local function creates a new declaration space called a ***local variable declaration space***. Names are introduced into this declaration space through parameters (*fixed_parameter*s and *parameter_array*s) and *type_parameter*s. The set and init accessor for a property or an indexer introduce the name `value` as a parameter. The body of the function member, anonymous function, or local function, if any, is considered to be nested within the local variable declaration space. When a local variable declaration space and a nested local variable declaration space contain elements with the same name, within the scope of the nested local name, the outer local name is hidden ([§7.7.1](basic-concepts.md#771-general)) by the nested local name.
```

RexJaeschke commented on standard/attributes.md:
```diff
@@ -105,9 +105,9 @@ class X : Attribute { ... }
 ### 23.2.3 Positional and named parameters
 
-Attribute classes can have ***positional parameter***s and ***named parameter***s. Each public instance constructor for an attribute class defines a valid sequence of positional parameters for that attribute class. Each non-static public read-write field and property for an attribute class defines a named parameter for the attribute class. For a property to define a named parameter, that property ...
+Attribute classes can have ***positional parameter***s and ***named parameter***s. Each public instance constructor for an attribute class defines a valid sequence of positional parameters for that attribute class. Each non-static public read-write field and non-static public read-write or read-init property for an attribute class defines a named parameter for the attribute class. For a property  ...
```
```suggestion
Attribute classes can have ***positional parameter***s and ***named parameter***s. Each public instance constructor for an attribute class defines a valid sequence of positional parameters for that attribute class. Each non-static public read-write field and non-static public read-write or read-init property for an attribute class defines a named parameter for the attribute class. For a property to define a named parameter, that property shall have both a public get accessor and a public set or init accessor.
```

RexJaeschke commented on standard/attributes.md:
```diff
@@ -465,7 +465,7 @@ The compilation of an *attribute* with attribute class `T`, *positional_argumen
 - For each *named_argument* `Arg` in `N`:
   - Let `Name` be the *identifier* of the *named_argument* `Arg`.
-  - `Name` shall identify a non-static read-write public field or property on `T`. If `T` has no such field or property, then a compile-time error occurs.
+  - `Name` shall identify a non-static read-write public field or a non-static public read-write or read-init property on `T`. If `T` has no such field or property, then a compile-time error occurs.
```
```suggestion
  - `Name` shall identify a non-static read-write public field or a public non-static read-write or read-init property on `T`. If `T` has no such field or property, then a compile-time error occurs.
```

RexJaeschke commented on standard/expressions.md:
```diff
@@ -450,27 +450,27 @@ Once a particular function member has been identified at binding-time, possibly
 >     <td rowspan="6">Property access</td>
 >     <td><code>P</code></td>
->     <td>The get accessor of the property <code>P</code> in the containing class or struct is invoked. A compile-time error occurs if <code>P</code> is write-only. If <code>P</code> is not <code>static</code>, the instance expression is <code>this</code>.</td>
+>     <td>The get accessor of the property <code>P</code> in the containing class or struct is invoked. A compile-time error occurs if <code>P</code> is write-only or init-only. If <code>P</code> is not <code>static</code>, the instance expression is <code>this</code>.</td>
```
Should "containing class or struct" be extended to include interface? Here and a few other table rows below?

RexJaeschke commented on standard/classes.md:
```diff
@@ -3373,14 +3377,18 @@ The differences between static and instance members are discussed further in [§
-*Note*: This subclause applies to both properties ([§15.7](classes.md#157-properties)) and indexers ([§15.9](classes.md#159-indexers)). The subclause is written in terms of properties, when reading for indexers substitute indexer/indexers for property/properties and consult the list of differences between properties and indexers given in [§15.9.2](classes.md#1592-indexer-and-property-differences) ...
+#### §accessors-general General
+
+*Note*: This subclause and its sibling subclauses apply to both properties ([§15.7](classes.md#157-properties)) and indexers ([§15.9](classes.md#159-indexers)). The clause is written in terms of properties, when reading for indexers substitute indexer/indexers for property/properties and consult the list of differences between properties and indexers given in [§15.9.2](classes.md#1592-indexer-and ...
```
```suggestion
*Note*: This subclause and its sibling subclauses apply to both properties ([§15.7](classes.md#157-properties)) and indexers ([§15.9](classes.md#159-indexers)). The text is written in terms of properties; when reading for indexers substitute indexer/indexers for property/properties and consult the list of differences between properties and indexers given in [§15.9.2](classes.md#1592-indexer-and-property-differences). *end note*
```

RexJaeschke commented on standard/classes.md:
```diff
@@ -3528,6 +3546,34 @@ Based on the presence or absence of the get and set accessors, a property is cla
+> *Example*: Consider the following, immutable type, which has auto-implemented properties:
+>
+> <!-- Example: {template:"standalone-lib", name:"AccessorsInit1"} -->
+> <!-- FIX: create and add file PointStructWithInit.cs. -->
```
```suggestion
```

RexJaeschke committed:
Update classes.md

RexJaeschke committed:
Update classes.md

RexJaeschke committed:
Apply suggestions from code review

I made these tweaks after comparing the edits I researched 2+ years ago and what I found in this new feature branch.

RexJaeschke committed:
Apply suggestions from code review

I made these tweaks after comparing the edits I researched 2+ years ago and what I found in this new feature branch.

RexJaeschke committed:
remove link
